lib/appsignal/finch.ex

```diff
@@ -0,0 +1,62 @@
+defmodule Appsignal.Finch do
+  require Appsignal.Utils
+
+  @tracer Appsignal.Utils.compile_env(:appsignal, :appsignal_tracer, Appsignal.Tracer)
+  @span Appsignal.Utils.compile_env(:appsignal, :appsignal_span, Appsignal.Span)
+  import Appsignal.Utils, only: [module_name: 1]
+
+  @moduledoc false
+
+  require Logger
+
+  def attach do
+    handlers = %{
+      [:finch, :request, :start] => &__MODULE__.finch_request_start/4,
+      [:finch, :request, :stop] => &__MODULE__.finch_request_stop/4
+    }
+
+    for {event, fun} <- handlers do
+      case :telemetry.attach({__MODULE__, event}, event, fun, :ok) do
+        :ok ->
+          _ = Appsignal.Logger.debug("Appsignal.Finch attached to #{inspect(event)}")
+
+          :ok
+
+        {:error, _} = error ->
+          Logger.warn("Appsignal.Finch not attached to #{inspect(event)}: #{inspect(error)}")
+
+          error
+      end
+    end
+  end
+
+  def finch_request_start(
+        _event,
+        _measurements,
+        %{name: name, request: request},
+        _config
+      ) do
+    do_finch_request_start(@tracer.current_span(), name, request)
+  end
+
+  defp do_finch_request_start(nil, _name, _request), do: nil
+
+  defp do_finch_request_start(parent, name, request) do
+    uri = %URI{
+      scheme: Atom.to_string(request.scheme),
+      path: request.path,
+      query: request.query,
```

Comment:
Do we want to add the query string to it? I'm not sure what we do for other integrations. It's probably okay, but I was thinking of old-school style APIs that use `?api_key=XXXX` instead of `Authorization` headers.

Reply:
We don't send the query along in other instrumentations, like Ruby's Net::HTTP instrumentation. I don't think we have to here. Neither do we need to store the path if we follow the example set by the Net::HTTP instrumentation.

Storing the query params (and path to some extend) makes events less easy to group by event category, name and body. Those three values create the digest and if that digest is unique for every event we lose any optimizations there.

Let's leave out the query and path.

Reply:
> I care about the `[:cool_library, :action]` events (which I know are created by `:telemetry.span`, so they follow the `:start`/`:stop`/`:exception` pattern) and I want to see the `:interesting_data` key of the metadata object.

I played around with this a bit based on this integration and the existing LiveView integration in the Phoenix integration. I can make this a bit smoother with some helpers which would remove some duplication. We can, for example remove some duplication in the attach function and the stop and exception handlers.

That said, I’m alright shipping this as-is, and revisiting this issue when we add another integration in this library.